beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,515 @@
+	id       enode.ID
+	isPinged bool
+	node     *enode.Node
+	peerID   peer.ID
+	topics   map[gossipsubcrawler.Topic]struct{}
+}
+
+type crawledPeers struct {
+	g *GossipsubPeerCrawler
+
+	mu       sync.RWMutex
+	byEnode  map[enode.ID]*peerNode
+	byPeerId map[peer.ID]*peerNode
+	byTopic  map[gossipsubcrawler.Topic]map[peer.ID]struct{}
+}
+
+func (cp *crawledPeers) updateStatusToPinged(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	existingPNode, ok := cp.byEnode[enodeID]
+	if !ok {
+		return
+	}
+
+	// we only want to ping a node with a given NodeId once -> not on every sequence number change
+	// as ping is simply a test of a node being reachable and not fake
+	existingPNode.isPinged = true
+}
+
+func (cp *crawledPeers) removePeerOnPingFailure(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	existingPNode, ok := cp.byEnode[enodeID]
+	if !ok {
+		return
+	}
+
+	// same idea as in "updateStatusToPinged" above.
+	// We don't want to test pings for every sequence number change for a given node as that
+	// can lead to an explosion in the number of pings the crawler needs to do.
+	// So, remove the peer when the first ping fails. If the node becomes reachable later,
+	// we will discover it during a re-crawl and ping it again to test for reachability.
+	// we're not blacklisting this peer anyways.
+	cp.updateTopicsUnlocked(existingPNode, nil)
+}
+
+func (cp *crawledPeers) updateCrawledIfNewer(node *enode.Node, topics []string) {
+	cp.mu.Lock()
+
+	enodeID := node.ID()
+	existingPNode, ok := cp.byEnode[enodeID]
+
+	if ok && existingPNode.node == nil {
+		log.WithField("enodeId", enodeID).Error("enode is nil for enodeId")
+		cp.mu.Unlock()
+		return
+	}
+
+	// we don't want to update enodes with a lower sequence number as they're stale records
+	if ok && existingPNode.node.Seq() >= node.Seq() {
+		cp.mu.Unlock()
+		return
+	}
+
+	if !ok {
+		// this is a new peer
+		peerID, err := enodeToPeerID(node)
+		if err != nil {
+			log.WithError(err).WithField("node", node.ID()).Debug("Failed to convert enode to peer ID")
+			cp.mu.Unlock()
+			return
+		}
+		existingPNode = &peerNode{
+			id:     enodeID,
+			node:   node,
+			peerID: peerID,
+			topics: make(map[gossipsubcrawler.Topic]struct{}),
+		}
+		cp.byEnode[enodeID] = existingPNode
+		cp.byPeerId[peerID] = existingPNode
+	} else {
+		existingPNode.node = node
+	}
+
+	cp.updateTopicsUnlocked(existingPNode, topics)
+
+	if existingPNode.isPinged || len(topics) == 0 {
+		cp.mu.Unlock()
+		return
+	}
+	cp.mu.Unlock()
+
+	select {
+	case cp.g.pingCh <- *node:
+	case <-cp.g.ctx.Done():
+		return
+	}
+}
+
+func (cp *crawledPeers) removeTopic(topic gossipsubcrawler.Topic) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	// Get all peers subscribed to this topic
+	peers, ok := cp.byTopic[topic]
+	if !ok {
+		return // Topic doesn't exist
+	}
+
+	// Remove the topic from each peer's topic list
+	for peerID := range peers {
+		if pnode, exists := cp.byPeerId[peerID]; exists {
```

Comment:
Nit: In general, we use `ok` instead of `exists`.

Reply:
Done.